src/NHSD.BuyingCatalogue.Identity.Api/Infrastructure/ValidationInputTagHelper.cs

```diff
@@ -0,0 +1,108 @@
+﻿using System.ComponentModel.DataAnnotations;
+using System.Linq;
+using System.Reflection;
+using Microsoft.AspNetCore.Mvc.Rendering;
+using Microsoft.AspNetCore.Mvc.ViewFeatures;
+using Microsoft.AspNetCore.Razor.TagHelpers;
+
+namespace NHSD.BuyingCatalogue.Identity.Api.Infrastructure
+{
+    [HtmlTargetElement(TagHelperConstants.Div, Attributes = TagHelperName)]
+    public sealed class ValidationInputTagHelper : TagHelper
+    {
+        public const string TagHelperName = "nhs-validation-input";
+        public const string InputDataTestIdName = "input-data-test-id";
+        public const string FieldDataTestIdName = "field-data-test-id";
+        public const string ErrorDataTestIdName = "error-data-test-id";
+
+        private readonly IHtmlGenerator _htmlGenerator;
+
+        public ValidationInputTagHelper(IHtmlGenerator htmlGenerator)
+        {
+            _htmlGenerator = htmlGenerator;
+        }
+
+        [HtmlAttributeNotBound]
+        [ViewContext]
+        public ViewContext ViewContext { get; set; }
+
+        [HtmlAttributeName(TagHelperConstants.For)]
+        public ModelExpression For { get; set; }
+
+        [HtmlAttributeName(FieldDataTestIdName)]
+        public string FieldDataTestId { get; set; }
+
+        [HtmlAttributeName(InputDataTestIdName)]
+        public string InputDataTestId { get; set; }
+
+        [HtmlAttributeName(ErrorDataTestIdName)]
+        public string ErrorDataTestId { get; set; }
+
+        public override void Process(TagHelperContext context, TagHelperOutput output)
+        {
+            output.ThrowIfNull();
+
+            output.Content.Clear();
+
+            var outerDivBuilder = new TagBuilder(TagHelperConstants.Div);
+
+            if (ViewContext.ViewData.ModelState.ContainsKey(For.Name))
+            {
+                outerDivBuilder.AddCssClass(TagHelperConstants.NhsFormGroupError);
+            }
+
+            outerDivBuilder.Attributes[TagHelperConstants.DataTestId] = FieldDataTestId ?? $"{For.Name}-field";
+
+            var formGroup = new TagBuilder(TagHelperConstants.Div);
+            formGroup.AddCssClass(TagHelperConstants.NhsFormGroup);
+
+            var labelBuilder = _htmlGenerator.GenerateLabel(
```

Comment:
I wonder if it would make this easier to read by refactoring each component into a get method e.g. GetLabelBuilder(), GetValidationBuilder().